### PR TITLE
preliminary rules/engine split

### PR DESCRIPTION
**Pull request by cwong-ocaml**
As per #3703 . This split was done mostly mechanically, taking build_system and all dependencies and putting them into `dune_engine`, renaming `dune` to `dune_rules`. Then, `open Dune_engine` was put at the top of all files in `dune_rules`, and finally all the relevant build errors (due to other parts of the codebase referring to `Dune`) were fixed.

There should be no behavioral changes; this is only a refactoring.

I think, if there are no major objections, we should merge this soon, and we can work on bikeshedding the rest (which files should be in which library, etc) afterwards.

Signed-off-by: Cameron Wong <cwong@janestreet.com>

### CODE REVIEW

File: src/dune_rules/assets.ml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+let jbuild_plugin_ml =
```
Comment (aalekseyev) [inline src/dune_rules/assets.ml:1]: This file looks duplicated now (it's both in engine and in rules).

**Review by aalekseyev — CHANGES_REQUESTED** [review]
Other than duplication of assets.ml, looks reasonable to me.

File: src/dune_rules/assets.ml (hunk 1 above)
Comment (cwong-ocaml) [reply src/dune_rules/assets.ml]: Hopefully fixed

**Review by aalekseyev — APPROVED** [review]

File: src/dune_rules/assets.ml [hunk 2]
```diff
@@ -1,4 +1,6 @@
-let jbuild_plugin_ml = {jbp|let () =
+let jbuild_plugin_ml =
+  {jbp|
+let () =
```
Comment (aalekseyev) [inline src/dune_rules/assets.ml:3]: You added an extra linebreak here. Why was this change needed?

Comment (cwong-ocaml) [reply src/dune_rules/assets.ml:3]: If it is not added, CI fails with a "code is not formatted properly" error.

Comment (aalekseyev) [reply src/dune_rules/assets.ml:3]: Hmm, but it wasn't there before...

Comment (cwong-ocaml) [reply src/dune_rules/assets.ml:3]: Yeah, I don't know what is or isn't going wrong. Basically all of the previous commits were failing CI with a formatting error, and running ocamlformat 0.14.3 on this file (alternatively, `make fmt`) also produces this change.

Comment (cwong-ocaml) [reply src/dune_rules/assets.ml:3]: Hmm, perhaps this file shouldn't be checked into CI at all if it is generated from the dune rules.

Comment (emillon) [reply src/dune_rules/assets.ml]: this file was referenced in `.ocamlformat-ignore`, I think you just need to update this file.

Comment (cwong-ocaml) [reply src/dune_rules/assets.ml]: aha, that would do it, how careless of me.

I stand by what I said earlier, however -- I think that it does not make sense to check in such a file to git when it is entirely a wrapper around another.

**Comment by rgrinberg** [comment]
> I stand by what I said earlier, however -- I think that it does not make sense to check in such a file to git when it is entirely a wrapper around another. However, I suppose such a change can be argued over later.

The generated file is checked in to simplify the bootstrap procedure. It's not simple to generate things when bootstrapping because we don't have the full build system available.
